### ISSUE TITLE
(req[i]) == (kWriteTo) in concat layer

### ISSUE THREAD

**Issue by SkidanovAlex** [dmlc/mxnet#1130]
I am building a Neural GPU from this paper: http://arxiv.org/pdf/1511.08228v1.pdf

I am facing an issue that I can't repro with a small demo. The code is here: https://github.com/SkidanovAlex/NeuralGPU/blob/49f4e83d46630132c67d470d5c96267ba1ddebcd/src/neuralgpu.py

When I run the code as it is right now, it runs. If I uncomment lines 15 and 25 and comment 24, the code fails when it binds with an error `(req[i]) == (kWriteTo)` when instantiating the ConcatLayer.

What I am trying to achieve is to pad my input with zeros. For that I create a new variable, which later will be initialized with zero, and concat my input with that variable (there's also BlockGrad there, which I assume is not needed, but removing it doesn't help). The error occurs when `i` is equal to `1`, and `req[i]` is equal to `kNullOp`. Am I missing some important step in initializing my `padding` variable?

Also, is there any better way to pad a tensor of shape `a x 1 x c x d` with zeros to shape `a x b x c x d` then concating it with a tensor of zeros?

And I have one more unrelated issue with the same code: all the weights of my model do not depend on the input size, yet the model itself does depend on it (some reshape layers use input size). I then want to train my model on inputs of different sizes sharing the weights. To achieve that I bind a model separately each time, and copy weights and biases from the last model. After several iterations I run out of GPU memory, so it seems like the parameters from the previous iterations are not freed up. Does mxnet free up the GPU memory as soon as the model symbol and all the parameters go out of scope?

**Comment by piiswrong** [comment]
You can make a new pad operator. I'm not sure what happened in your case. Could you print out req[i] and check what it is?

**Comment by SkidanovAlex** [comment]
@piiswrong as I mentioned, `i` is `1` and `req[i]` is `kNullOp`. I can get around it by multiplying `data` by zero (effectively getting a vector of zeros), and get the behavior needed. But I am quite convinced that this is a bug in `mxnet`, because everything I do for the `input` variable I also do for `padding`, yet passing `padding` as the second argument leads to this issue, while passing something that is derived from `input` works.

Can you please comment on my second issue: does `mxnet` free up GPU memory as soon as model goes out of scope?

**Comment by piiswrong** [comment]
Ok. This is because pad doesn't require gradient. This should be an easy fix. In concat-inl.h: Backward, change CHECK_EQ(req[i], kWriteTo) to CHECK(req[i] ==  kWriteTo || req[i] ==  kNullOp) and when it's NullOp, don't propagate down the gradient.

**Comment by piiswrong** [comment]
It frees up memory when NDArrays go out of scope. You need to wait for operations from previous iter to finish (because they are pushed to engine queue async) before binding the next model. Try for arr in your_weights: arr.wati_to_read()

**Comment by tqchen** [comment]
Normally it should be OK, as all the destructions are also scheduled, so things won't be freed until the related ops are completed

**Comment by piiswrong** [comment]
@tqchen The problem is executor.__del__ returns before it's destruction is really executed. The next round will start pushing ops to the engine before the previous one finish. This makes manual synchronization necessary.

**Comment by tqchen** [comment]
The del calls the destructor, whose destruction will be pushed to the engine, then next round ops are also pushed, and will follow the previous destruction if there is any dependency

**Comment by SkidanovAlex** [comment]
I will send a PR for the (req) == (kWriteTo) as soon as I finish a test for it, in couple days.

Wrt the freeing issue, I added some tracing and currently observe that in my particular case the memory is never deallocated. Most likely I accidentally leak refs to my ndarrays, I will post an update here if I narrow it down to an issue in mxnet.

**Comment by SkidanovAlex** [comment]
It seems to me that there's some kind of caching in place. Consider this example. ALLOCING is traced every time when `cudaMalloc` is called. `cudaFree` also traces when it is called, so it appears to be never called in this example:

```
>>> ctx = mx.gpu(0)
>>> def test():
...     mx.nd.zeros((100, 100), ctx)
...
>>> test()
[11:35:53] src/storage/./gpu_device_storage.h:40: ALLOCING 40000 0x700c09e00
>>> test()
>>> test()
```

Notice how after first  call to `test` the array is allocated, but not freed, even though it goes out of scope as soon as the function ends. However, when I call the function again, it doesn't allocate anything, so it reused the previously allocated memory. If I replace `zeros` with `mx.nd.array`, the behavior is the same.

Am I right that the chunks of memory are cached in some way? Can I disable this behavior? In my case I call the same function on different shapes, and therefore the cached memory is not reused and wasted, resulting in `Out of Memory` error

**Comment by piiswrong** [comment]
@tqchen I know this consistent, but it's causing OOM

**Comment by SkidanovAlex** [comment]
Ok, I narrowed down the problem. The problem is in storage.cc. It uses `PooledStorageManager`, which, in fact, does do caching. It has some logic to prevent OOM, but there are two issues with it:

1. The threshold is hardcoded to 4GB. It is hardcoded to an extent that it is a template argument, not even a variable :) My videocard has 2GB of memory, so obviously it OOMs way before the threshold is reached.

2. The threshold check is broken anyway. Look at this code from `pooled_storage_manager.h`:

```
    if (kThreshold <= used_memory_) {
      ReleaseAll();
    }
    used_memory_ += size;
    return DeviceStorage::Alloc(size);
```

It compares used memory with the threshold before it increases the memory used, so if an allocation exceeds 4GB, it will go through, and cause OOM even if I do have 4GB. The logic should be:

```
    used_memory_ += size;
    if (kThreshold <= used_memory_) {
      ReleaseAll();
    }
    return DeviceStorage::Alloc(size);
```

For a fix I suggest to make the threshold a variable instead of a template argument, and set it to the total GPU memory. I would also suggest to add an environment variable that allows changing it.

For my purposes for now I just reduced the threshold to 1.5GB in the source code.
